test/jdk/java/nio/channels/DatagramChannel/SelectWhenRefused.java

```diff
@@ -99,4 +106,22 @@ static void sendDatagram(DatagramChannel dc, SocketAddress remote)
         ByteBuffer bb = ByteBuffer.wrap("Greetings!".getBytes());
         dc.send(bb, remote);
     }
+
+    static void logUnexpectedWakeup(DatagramChannel dc, Set<SelectionKey> selectedKeys) {
+        for (SelectionKey key : selectedKeys) {
+            if (!key.isValid() || !key.isReadable()) {
+                System.out.println("Invalid or unreadable key: " + key);
+                return;
+            }
+            try {
+                System.out.println("Attempting to read datagram from key: " + key);
+                ByteBuffer buf = ByteBuffer.allocate(100);
+                SocketAddress sa = dc.receive(buf);
```

Comment:
The datagram channel is configured non-blocking and there may not be a datagram to receive, therefore sa may be null and has to be checked here.

Reply:
If we reach here the key is both valid and readable so shouldn't there be something to receive, even if the channel is non-blocking?